packages/flutter_tools/lib/src/android/migrations/min_sdk_version_migration.dart

```diff
@@ -0,0 +1,47 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../../base/project_migrator.dart';
+import '../../project.dart';
+
+@visibleForTesting
+const String minSdk16 = 'minSdkVersion 16';
+@visibleForTesting
+const String minSdk17 = 'minSdkVersion 17';
+@visibleForTesting
+const String minSdk18 = 'minSdkVersion 18';
+@visibleForTesting
+const String flutterMinSdk = 'minSdkVersion flutter.minSdkVersion';
+@visibleForTesting
+const String appGradleNotFoundWarning = 'Module level build.gradle file not found, skipping minSdkVersion migration.';
+
+class MinSdkVersionMigration extends ProjectMigrator {
+  MinSdkVersionMigration(
+      AndroidProject project,
+      super.logger,
+  ) : _project = project;
+
+
+  final AndroidProject _project;
+
+  @override
+  void migrate() {
+    // Skip if we cannot find the app level build.gradle file.
+    if (!_project.appGradleFile.existsSync()) {
+      logger.printTrace(appGradleNotFoundWarning);
+      return;
+    }
+
+    processFileLines(_project.appGradleFile);
+  }
+
+  @override
+  String migrateFileContents(String fileContents) {
+    return fileContents.replaceAll(minSdk16, flutterMinSdk)
```

Comment:
Open to thoughts on if we should take a different approach for replacing. This was inspired by the PR that updates the [minimum iOS version](https://github.com/jmagman/flutter/blob/84a244b3d57449f4b50c74520db2b804c490f000/packages/flutter_tools/lib/src/ios/migrations/deployment_target_migration.dart#L60), but could be replaced with a regex or other method.

Reply:
You could rewrite this as a regular expression like 'minSdkVersion [1-9][0-9]+' but that might end up catching future version numbers that we don't want to migrate. It would also match (and thus replace) cases where minSdkVersion is > flutterMinSdk.

I would leave this as is.

Reply:
I think it'd be reasonable to write this as a regular expression that matches exactly the same set of versions that the three strings match: `minSdkVersion 1[678]`. I agree it shouldn't use a pattern that matches later versions we don't intend to migrate.

---

A further step that may be useful, and which a regular expression enables, is to make this a bit more robust against accidentally rewriting matches that might occur in comments or strings. Also against lines like `minSdkVersion 1600` which have one of these as a substring but mean something different.

That class of issue is less of a concern for the iOS migration: it's applying to an XML file (so the string to replace has built-in delimiters like `<key>`/`</key>`), plus it's an XML file that is made of simple data mostly edited by machine. Here, these Gradle scripts contain a fair bit of logic, and humans do edit them, so comments are natural.

Here's a regexp that takes that further step — it avoids rewriting lines that have anything else on them (except a possible trailing comment), and therefore mean something other than what we're intending to migrate:
```dart
final RegExp _oldMinSdkVersionRe = RegExp(r'(?<=^\s*)minSdkVersion 1[678](?=\s*(?://|$))', multiLine: true);
```

Here's a test that exercises the comments case, and fails with the plain-string-pattern version but passes with that regexp:
```dart
      testWithoutContext('avoid rewriting comments', () {
        const String code = '// minSdkVersion 16  // old default\n'
                    '        minSdkVersion 23  // new version';
        project.appGradleFile.writeAsStringSync(sampleModuleGradleBuildFile(code));
        migration.migrate();
        expect(project.appGradleFile.readAsStringSync(), sampleModuleGradleBuildFile(code));
      });
```

Reply:
For additional context (that I should have included in the PR): @reidbaker and I had talked about this offline and decided initially that we were ok taking the simple approach of find and replace, despite it hitting comments, but this argument:
> That class of issue is less of a concern for the iOS migration: it's applying to an XML file (so the string to replace has built-in delimiters like <key>/</key>), plus it's an XML file that is made of simple data mostly edited by machine. Here, these Gradle scripts contain a fair bit of logic, and humans do edit them, so comments are natural.

convinces me that taking a regular expression approach similar to the java-gradle migrator is the better way. I'll change to include the regexp and comment-test-case